### PR TITLE
Closes #2141 - Fixes Parquet SegArray Read Index Bug

### LINKED ISSUE #2141: Parquet SegArray Read Index Bug
This morning it was reported that there was an issue with reading SegArray objects from Parquet files on distributed systems.

After doing some digging, it was determined that this was due to the domains being improperly assigned.

### PR DESCRIPTION
**Pull request by Ethan-DeBandi99**
Closes #2141 

Corrects read indexing to account for distributed or singular files. Adds test case for distributed files.

*Commented out code in test is needed for #2145, which I will be working next. On that same note, this implementation is not fully functional for SegArrays containing empty segments being read.*

### CODE REVIEW

File: src/ParquetMsg.chpl [hunk 1]
```diff
@@ -181,26 +181,29 @@ module ParquetMsg {
   proc readListFilesByName(A: [] ?t, filenames: [] string, sizes: [] int, dsetname: string, ty) throws {
     extern proc c_readListColumnByName(filename, chpl_arr, colNum, numElems, startIdx, batchSize, errMsg): int;
     var (subdoms, length) = getSubdomains(sizes);
+    var fileOffsets = (+ scan sizes) - sizes;
+    writeln("\n\nFilenames: %jt\nSizes: %jt".format(filenames, sizes));
```
Comment (pierce314159) [inline src/ParquetMsg.chpl:185]: is this a left over print statement

File: src/ParquetMsg.chpl [hunk 2]
```diff
@@ -181,26 +181,29 @@ module ParquetMsg {
...
       var locFiles = filenames;
       var locFiledoms = subdoms;
+      var locOffsets = fileOffsets;
 
       try {
-        forall (filedom, filename) in zip(locFiledoms, locFiles) {
+        forall (off, filedom, filename) in zip(locOffsets, locFiledoms, locFiles) {
           for locdom in A.localSubdomains() {
             const intersection = domain_intersection(locdom, filedom);
+            writeln("\n\nFileDom: %jt\nLocDom: %jt\nIntersection: %jt\nOffset: %i".format(filedom, locdom, intersection, off));
```
Comment (pierce314159) [inline src/ParquetMsg.chpl:196]: is this a left over print statement

File: src/ParquetMsg.chpl [hunk 3]
```diff
@@ -181,26 +181,29 @@ module ParquetMsg {
...
 
-              if c_readListColumnByName(filename.localize().c_str(), c_ptrTo(col),
-                                    dsetname.localize().c_str(), intersection.size, 0,
+              if c_readListColumnByName(filename.localize().c_str(), c_ptrTo(A[intersection.low]),
+                                    dsetname.localize().c_str(), intersection.size, intersection.low - off,
                                     batchSize, c_ptrTo(pqErr.errMsg)) == ARROWERROR {
                 pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
               }
-              A[filedom] = col;
+              // A[intersection] = col;
```
Comment (pierce314159) [inline src/ParquetMsg.chpl:206]: should we rm this commented out code?

File: src/ParquetMsg.chpl [hunk 4]
```diff
@@ -326,9 +330,12 @@ module ParquetMsg {
   proc getArrSize(filename: string) throws {
     extern proc c_getNumRows(chpl_str, errMsg): int;
     var pqErr = new parquetErrorMsg();
+
+    
     
     var size = c_getNumRows(filename.localize().c_str(),
                             c_ptrTo(pqErr.errMsg));
+    writeln("\n\nFile %s has %i rows".format(filename, size));
```
Comment (pierce314159) [inline src/ParquetMsg.chpl:338]: is this a left over print statement

File: src/ParquetMsg.chpl [hunk 5]
```diff
@@ -645,6 +652,9 @@ module ParquetMsg {
     var filedom = filenames.domain;
     var segments = makeDistArray(len, int);
     var listSizes: [filedom] int = calcListSizesandOffset(segments, filenames, sizes, dsetname);
+    writeln("Segments-PRE: %jt".format(segments));
+    segments = (+ scan segments) - segments;
+    writeln("Segments: %jt".format(segments));
```
Comment (pierce314159) [inline src/ParquetMsg.chpl:657]: are these left over print statements

**Review by pierce314159 — APPROVED** [review]
Looks good to me! I had an idea for the empty segments case which i left in the issue

File: src/ArrowFunctions.cpp [hunk 6]
```diff
@@ -456,6 +532,7 @@ int64_t cpp_getStringColumnNullIndices(const char* filename, const char* colname
 
 int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* colname, int64_t numElems, int64_t startIdx, int64_t batchSize, char** errMsg) {
   try {
+    std::cout << "\n\nStartIdx: " << startIdx << "NumElems" << numElems << "\n\n";
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply src/ArrowFunctions.cpp:535]: yeah. Removed now.

File: src/ParquetMsg.chpl (hunk 1 above)
Comment (Ethan-DeBandi99) [reply src/ParquetMsg.chpl:185]: Removed

File: src/ParquetMsg.chpl (hunk 2 above)
Comment (Ethan-DeBandi99) [reply src/ParquetMsg.chpl:196]: Removed

File: src/ParquetMsg.chpl (hunk 3 above)
Comment (Ethan-DeBandi99) [reply src/ParquetMsg.chpl:206]: removed

File: src/ParquetMsg.chpl (hunk 4 above)
Comment (Ethan-DeBandi99) [reply src/ParquetMsg.chpl:338]: removed

File: src/ParquetMsg.chpl (hunk 5 above)
Comment (Ethan-DeBandi99) [reply src/ParquetMsg.chpl:657]: removed

**Pull request merged**